File: leetcode/python/0110. Balanced Binary Tree.py
# Time complexity: O(N), where N is the number of nodes
# Space complexity: O(N)

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True
        if abs(self.getHeight(root.left) - self.getHeight(root.right)) <= 1:
            return self.isBalanced(root.left) and self.isBalanced(root.right)
        return False

    def getHeight(self, node):
        if node is None:
            return -1
        elif node.left == None and node.right == None:
            return 0
        else:
            return max(self.getHeight(node.left), self.getHeight(node.right)) + 1

File: leetcode/java/0110. Balanced Binary Tree.java
package leetcode.java;

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// Time complexity: O(N), where N is the number of nodes
// Space complexity: O(N)

class Solution {
    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }
        if (Math.abs(this.getHeight(root.left) - this.getHeight(root.right)) <= 1) {
            return this.isBalanced(root.left) && this.isBalanced(root.right);
        }
        return false;
    }

    public int getHeight(TreeNode root) {
        if (root == null) {
            return -1;
        }
        return 1 + Math.max(this.getHeight(root.left), this.getHeight(root.right));
    }
}